securedrop/journalist_app/sessions.py

```diff
@@ -0,0 +1,278 @@
+        self.flash = flash
+        self.uid = None
+        self.user = None
+        self.to_destroy = True
+
+    def regenerate(self) -> None:
+        self.to_regenerate = True
+
+
+class SessionInterface(FlaskSessionInterface):
+
+    def _generate_sid(self) -> str:
+        return token_urlsafe(32)
+
+    def _get_signer(self, app: Flask) -> URLSafeTimedSerializer:
+        if not app.secret_key:
+            raise RuntimeError('No secret key set')
+        return URLSafeTimedSerializer(app.secret_key,
+                                      salt=self.salt)
+
+    """Uses the Redis key-value store as a session backend.
+
+    :param redis: A ``redis.Redis`` instance.
+    :param key_prefix: A prefix that is added to all Redis store keys.
+    :param salt: Allows to set the signer salt from the calling interface
+    :param header_name: if use_header, set the header name to parse
+    """
+
+    def __init__(self, lifetime: int, renew_count: int, redis: Optional[Redis],
+                 key_prefix: str, salt: str, header_name: str) -> None:
+        self.serializer = session_json_serializer
+        if redis is None:
+            redis = Redis()
+        self.redis = redis
+        self.lifetime = lifetime
+        self.renew_count = renew_count
+        self.key_prefix = key_prefix
+        self.api_key_prefix = 'api_' + key_prefix
+        self.salt = salt
+        self.api_salt = 'api_' + salt
+        self.header_name = header_name
+        self.new = False
+        self.has_same_site_capability = hasattr(self, "get_cookie_samesite")
+
+    def _new_session(self, is_api: bool = False, initial: Any = None) -> ServerSideSession:
+        sid = self._generate_sid()
+        token: str = self._get_signer(app).dumps(sid)  # type: ignore
+        session = ServerSideSession(sid=sid,
+                                    token=token,
+                                    lifetime=self.lifetime,
+                                    initial=initial)
+        session.new = True
+        session.is_api = is_api
+        return session
+
+    def open_session(self, app: Flask, request: Request) -> Optional[ServerSideSession]:
+        '''This function is called by the flask session interface at the
+        beginning of each request.
+        '''
+        is_api = (request.path.split('/')[1] == 'api')
+
+        if is_api:
+            self.key_prefix = self.api_key_prefix
+            self.salt = self.api_salt
+            auth_header = request.headers.get(self.header_name)
+            if auth_header:
+                split = auth_header.split(" ")
+                if len(split) != 2 or split[0] != 'Token':
+                    return self._new_session(is_api)
+                sid: Optional[str] = split[1]
+            else:
+                return self._new_session(is_api)
+        else:
+            sid = request.cookies.get(app.session_cookie_name)
+        if sid:
+            try:
+                sid = self._get_signer(app).loads(sid)
```

Comment:
Just to confirm: the max_age argument does not need to be supplied to the loads() call here, to reject expired signatures?

Reply:
In my opinion it would not add any security as the session gets destroyed server side on expiration and thus regardless of the signature the signed session identifier in the cookie does not point to anything meaningful anymore. Also there are negligible chances for session identifiers to be reused ever.